src/main/java/net/fabricmc/fabric/impl/client/render/FluidRenderHandlerRegistryImpl.java

```diff
@@ -0,0 +1,85 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.client.render;
+
+import net.fabricmc.fabric.api.client.render.FluidRenderHandler;
+import net.fabricmc.fabric.api.client.render.FluidRenderHandlerRegistry;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.render.block.BiomeColors;
+import net.minecraft.client.render.block.FluidRenderer;
+import net.minecraft.client.texture.Sprite;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.ExtendedBlockView;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+public class FluidRenderHandlerRegistryImpl implements FluidRenderHandlerRegistry {
+    public static final FluidRenderHandlerRegistryImpl INSTANCE = new FluidRenderHandlerRegistryImpl();
+    private final Map<Fluid, FluidRenderHandler> handlers = new IdentityHashMap<>();
+    private final Map<Fluid, FluidRenderHandler> modHandlers = new IdentityHashMap<>();
+
+    private FluidRenderHandlerRegistryImpl() {
+
+    }
+
+    @Override
+    public FluidRenderHandler get(Fluid fluid) {
+        return handlers.get(fluid);
+    }
+
+    public FluidRenderHandler getModded(Fluid fluid) {
+        return modHandlers.get(fluid);
+    }
+
+    @Override
+    public void register(Fluid fluid, FluidRenderHandler renderer) {
+        handlers.put(fluid, renderer);
+        modHandlers.put(fluid, renderer);
+    }
+
+    public void onFluidRendererReload(Sprite[] waterSprites, Sprite[] lavaSprites) {
+        FluidRenderHandler waterHandler = new FluidRenderHandler() {
+            @Override
+            public Sprite[] getFluidSprites(ExtendedBlockView view, BlockPos pos, FluidState state) {
+                return waterSprites;
+            }
+
+            @Override
+            public int getFluidColor(ExtendedBlockView view, BlockPos pos, FluidState state) {
+                return BiomeColors.waterColorAt(view, pos);
+            }
+        };
+
+        //noinspection Convert2Lambda
+        FluidRenderHandler lavaHandler = new FluidRenderHandler() {
+            @Override
+            public Sprite[] getFluidSprites(ExtendedBlockView view, BlockPos pos, FluidState state) {
+                return lavaSprites;
+            }
+        };
+
+        register(Fluids.WATER, waterHandler);
+        register(Fluids.FLOWING_WATER, waterHandler);
+        register(Fluids.LAVA, lavaHandler);
+        register(Fluids.FLOWING_LAVA, lavaHandler);
+        handlers.putAll(modHandlers);
```

Comment:
This seems like the opportune time to cache all the sprites from modded handlers.

Reply:
Only if they can be assumed to be static, though...? Otherwise, it can't really *cache* them, nicely.

Reply:
Well, it they aren't static that's going to be a performance problem, no?  I think that should be required unless there's a custom tessellator. In that case we don't care.